src/Symfony/Component/DependencyInjection/Container.php

```diff
@@ -206,15 +206,29 @@ public function has($id)
             ) {
                 return true;
             }
+
+            if (isset($this->privates[$id])) {
+                @trigger_error(sprintf('Checking for the existence of the "%s" private service is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0.', $id), E_USER_DEPRECATED);
+            }
+
+            if (isset($this->methodMap[$id])) {
+                return true;
+            }
+
             if (--$i && $id !== $lcId = strtolower($id)) {
                 $id = $lcId;
-            } else {
-                if (isset($this->privates[$id])) {
-                    @trigger_error(sprintf('Checking for the existence of the "%s" private service is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0.', $id), E_USER_DEPRECATED);
-                }
+                continue;
+            }
+
+            // We only check the convention-based factory in a compiled container (i.e. a child class other than a ContainerBuilder,
+            // and only when the dumper has not generated the method map (otherwise the method map is considered to be fully populated by the dumper)
+            if (!$this->methodMap && !$this instanceof ContainerBuilder && __CLASS__ !== static::class && method_exists($this, 'get'.strtr($id, $this->underscoreMap).'Service')) {
```

Comment:
@nicolas-grekas in get, we could. But in has, this would not allow to warn properly in all cases

@fabpot the convention is the deprecated one. Adding a new interface won't help, as non-updated dumpers would not implement this interface either, and updated dumpers will never enter this code due to ``!$this->methodMap``. This code is precisely about handling things for custom dumpers to warn them to upgrade.